### PR TITLE
ref: django 3.x on_delete must be callable

### PR DESCRIPTION
**Pull request by asottile-sentry**

### CODE REVIEW

File: src/sentry/db/models/fields/hybrid_cloud_foreign_key.py [hunk 1]
```diff
@@ -45,21 +45,25 @@
...
 __all__ = ("HybridCloudForeignKey",)
 
-from typing import Any, Tuple
-
-from django.apps import apps
-
 
 class HybridCloudForeignKeyCascadeBehavior(IntEnum):
     CASCADE = 1
     SET_NULL = 2
```
Comment (asottile-sentry) [inline src/sentry/db/models/fields/hybrid_cloud_foreign_key.py:58]: @getsentry/hybrid-cloud this enum is unused outside this file -- should it be removed?

and/or can you make sure the change to this file is ok -- it is the only manual change of this PR

**Review by dashed — APPROVED** [review]

File: src/sentry/data_export/models.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ class ExportedData(Model):
     __include_in_export__ = False
 
     organization = FlexibleForeignKey("sentry.Organization")
-    user_id = HybridCloudForeignKey(settings.AUTH_USER_MODEL, null=True, on_delete="SET_NULL")
+    user_id = HybridCloudForeignKey(settings.AUTH_USER_MODEL, null=True, on_delete=models.SET_NULL)
```
Comment (corps) [inline src/sentry/data_export/models.py:36]: This seems dangerous.  This behavior does not match at all.
For instance, `models.SET_NULL` is a synchronous behavior that maps to a transaction completing.  However, "SET_NULL" is a HybridCloudForeignKey specific behavior that does not map to any of the `models.SET_NULL` guarantees.  I worry this is highly misleading.  I'm willing to change the name of the values, but I think it's dangerous to share the enum.

Comment (asottile-sentry) [reply src/sentry/data_export/models.py:36]: django 3.x does not allow this to be **stringly-typed** -- what would you suggest

File: src/sentry/db/models/fields/hybrid_cloud_foreign_key.py [hunk 3]
```diff
@@ -74,13 +78,19 @@ def foreign_table_name(self) -> str:
         return self.foreign_model._meta.db_table
 
     def __init__(
-        self, foreign_model: str, *, on_delete: HybridCloudForeignKeyCascadeBehavior | str, **kwds
+        self,
+        foreign_model: str,
+        *,
+        on_delete: HybridCloudForeignKeyCascadeBehavior | str | Callable[..., None],
```
Comment (corps) [inline src/sentry/db/models/fields/hybrid_cloud_foreign_key.py:84]: Ahhh, ok, I see the typing issue at its core.
This is fine.

File: src/sentry/data_export/models.py (hunk 2 above)
Comment (corps) [reply src/sentry/data_export/models.py:36]: Understood.  I'll take a stab at the PR.

I'm hoping that we can use a different name while retaining the same typing.  This way we can separate semantics, play cool with typing, and avoid a database state issue.

**Comment by asottile-sentry** [comment]
actually it's worse than that -- any of the non-HC ones with strings are silently doing nothing.

will update those in a separate PR

**Pull request closed without merging**